src/ping.c

```diff
@@ -346,67 +386,82 @@ static void *ping_thread(void *arg) /* {{{ */
   return (void *)0;
 } /* }}} void *ping_thread */
 
-static int start_thread(void) /* {{{ */
+
+static int stop_thread(void) /* {{{ */
 {
-  int status;
+  int status = 0;
 
   pthread_mutex_lock(&ping_lock);
 
-  if (ping_thread_loop != 0) {
-    pthread_mutex_unlock(&ping_lock);
-    return 0;
-  }
-
-  ping_thread_loop = 1;
-  ping_thread_error = 0;
-  status = plugin_thread_create(&ping_thread_id, ping_thread,
-                                /* arg = */ (void *)0, "ping");
-  if (status != 0) {
-    ping_thread_loop = 0;
-    ERROR("ping plugin: Starting thread failed.");
+  if (ping_thread_loop == 0) {
     pthread_mutex_unlock(&ping_lock);
     return -1;
   }
 
+  ping_thread_loop = 0;
+  pthread_cond_broadcast(&ping_cond);
   pthread_mutex_unlock(&ping_lock);
-  return 0;
-} /* }}} int start_thread */
 
-static int stop_thread(void) /* {{{ */
+  pthread_mutex_lock(&ping_lock);
+  for (sourcelist_t *sl = sourcelist_head; sl != NULL; sl = sl->next) {
+    if (sl->ping_thread_id <= 0) {
+      status = pthread_join(sl->ping_thread_id, /* return = */ NULL);
+      if (status != 0) {
+        ERROR("ping plugin: Stopping thread failed.");
+        status = -1;
```

Comment:
This may not work as intended: in the next iteration, `status` will be reset to the return value of `pthread_join`, which may be successful. So this function may return zero even though one of the joins failed.

I suggest to introduce a new variable, `int last_err = 0;` before the `for` loop and add `last_err = status;` to the error handling inside the loop.